### PR TITLE
Extension tray menu items

### PR DESCRIPTION
**Pull request by jweak**
Added API for extension to add menu items to tray menu.

https://user-images.githubusercontent.com/774344/147666598-cb848465-8c55-48de-bccf-b7037b043909.mov

### CODE REVIEW

File: src/main/tray/tray.ts [hunk 1]
```diff
@@ -62,7 +67,7 @@ export function initTray(windowManager: WindowManager) {
   const disposers = [
     autorun(() => {
       try {
-        const menu = createTrayMenu(windowManager);
+        const menu = createTrayMenu(windowManager, toJS(trayMenuItems.get()));
```
Comment (ixrock) [inline src/main/tray/tray.ts:70]: Why do we need `toJS()` here? As `trayMenuItems` is observable box afaik and calling `.get()` already unwraps all observability == getting plain js/json? Wdyt?

File: src/main/tray/tray.ts [hunk 2]
```diff
@@ -108,6 +116,8 @@ function createTrayMenu(windowManager: WindowManager): Menu {
     });
   }
 
+  template = template.concat(extensionTrayItems);
```
Comment (ixrock) [inline src/main/tray/tray.ts]: _nit:_
```suggestion
  template = [template, ...extensionTrayItems];
```

**Review by ixrock — DISMISSED** [review]
LGTM

File: src/main/tray/tray.ts (hunk 1 above)
Comment (jweak) [reply src/main/tray/tray.ts:70]: I created an extension that had `trayMenus` as an observable array and added a menu after the extension was loaded to test if the extension can change its menus and this `autorun` then went into infinite loop after the `trayMenus` changed. This fixed it. Not sure why, I'm not a `mobx` pro but maybe something to do with nested arrays?

File: src/main/menu/electron-menu-items.injectable.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ const electronMenuItemsInjectable = getInjectable({
     const extensions = di.inject(mainExtensionsInjectable);
 
     return computed(() =>
-      extensions.get().flatMap((extension) => extension.appMenus),
+      extensions.get().flatMap((extension) => extension.appMenus.slice()),
```
Comment (jansav) [inline src/main/menu/electron-menu-items.injectable.ts]: What is the reasoning behind `.slice()`?

Comment (jweak) [reply src/main/menu/electron-menu-items.injectable.ts]: It was for supporting observable arrays but it's not actually needed. good catch!

**Review by jansav — APPROVED** [review]
Looks good. Love the amount of new tests.

**Review by ixrock — DISMISSED** [review]

File: src/extensions/lens-main-extension.ts [hunk 4]
```diff
@@ -25,9 +25,11 @@ import { catalogEntityRegistry } from "../main/catalog";
 import type { CatalogEntity } from "../common/catalog";
 import type { IObservableArray } from "mobx";
 import type { MenuRegistration } from "../main/menu/menu-registration";
+import type { MenuItemConstructorOptions } from "electron";
```
Comment (jansav) [inline src/extensions/lens-main-extension.ts]: I would prefer if we wouldn't expose interfaces (or anything..) from 3rd party libraries in extension API. This limits our ability to adapt to future needs. 

3rd party libraries should always be isolated behind abstractions. 

In this case, we should create own interface which exposes only the features that we want to expose and then in `createTrayMenu` we should map it to the interface that electron requires. That gives us flexibility to do what ever we want in future, even if we choose to implement this menu in some other way.

**Review by jansav — CHANGES_REQUESTED** [review]

File: src/extensions/lens-main-extension.ts (hunk 4 above)
Comment (jweak) [reply src/extensions/lens-main-extension.ts]: Yeah I thought about. Fair enough. Added a type `TrayMenuRegistration` for that.

**Review by aleksfront — DISMISSED** [review]
Tested on both linux and win, works good!

**Review by ixrock — DISMISSED** [review]

File: docs/extensions/guides/main-extension.md [hunk 5]
```diff
@@ -96,6 +95,23 @@ export default class SamplePageMainExtension extends Main.LensExtension {
 When the menu item is clicked the `navigate()` method looks for and displays a global page with id `"myGlobalPage"`.
 This page would be defined in your extension's `Renderer.LensExtension` implementation (See [`Renderer.LensExtension`](renderer-extension.md)).
 
+### `trayMenus`
+
+`trayMenus` is a subset of Electron's [`MenuItemConstructorOptions`](https://www.electronjs.org/docs/v14-x-y/api/menu-item). Most importantly you can define a `label` and a `click` handler. Other properties are `submenu`, `enabled`, `visible`, `toolTip`, `id` and `type`.
```
Comment (jansav) [inline docs/extensions/guides/main-extension.md]: This is outdated

**Review by jansav — CHANGES_REQUESTED** [review]

**Review by jansav — DISMISSED** [review]

**Review by chenhunghan — APPROVED** [review]
LGTM

**Review by jansav — APPROVED** [review]

**Pull request merged**